### ISSUE TITLE
Remove hidden assumptions in BlockMatrix

### ISSUE THREAD

**Issue by saidctb** [pyccel/psydac#245]
BlockMatrix has some hidden assumptions, such as all its block are of the `StencilMatrix` type, especially when we generate the dot product and the transpose methods, this should be changed and possibly create another type to handle such case

**Comment by yguclu** [comment]
@saidctb @tomcaruso Do we still generate functions for `BlockMatrix.dot()` and `BlockMatrix.transpose()`? If not, we can close this issue.

**Comment by yguclu** [comment]
@jowezarek, @saidctb: Is this issue still relevant?

**Comment by jowezarek** [comment]
@yguclu I don't think any functions are generated for `BlockLinearOperator.dot()` or `BlockLinearOperator.transpose()` and I don't think we still assume the blocks of a `BlockLinearOperator` must be `StencilMatrix` objects. So imo this issue is no longer relevant.

**Comment by yguclu** [comment]
> @yguclu I don't think any functions are generated for `BlockLinearOperator.dot()` or `BlockLinearOperator.transpose()` and I don't think we still assume the blocks of a `BlockLinearOperator` must be `StencilMatrix` objects. So imo this issue is no longer relevant.

I agree that a `BlockLinearOperator` can now contain blocks of any subclass of `LinearOperator`.

Nevertheless, when we call `discretize` on a `BilinearForm` which corresponds to a system of equations, a special `dot` method used to be generated in this case, and it assumed that each block was a `StencilMatrix` object.

In fact, the method `allocate_matrices` of class `DiscreteBilinearForm` [calls](https://github.com/pyccel/psydac/blob/18758bc871e4c0cc411a72081fe5839538f996c6/psydac/api/fem.py#L860-L865) the method `set_backend` of class `BlockLinearOperator`, which in turn [calls](https://github.com/pyccel/psydac/blob/18758bc871e4c0cc411a72081fe5839538f996c6/psydac/linalg/block.py#L1306-L1397) `LinearOperatorDot`. The latter class is in charge of code generation.
